src/pools/abstract-pool.ts

```diff
@@ -0,0 +1,173 @@
+}
+
+interface IWorker {
+  on(event: 'error', handler: ErrorHandler<this>): void
+  on(event: 'online', handler: OnlineHandler<this>): void
+  on(event: 'exit', handler: ExitHandler<this>): void
+}
+
+class PoolEmitter extends EventEmitter {}
+
+export abstract class AbstractPool<
+  Worker extends IWorker,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  Data = any,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  Response = any
+> {
+  public readonly workers: Worker[] = []
+  public nextWorker: number = 0
+
+  /**
+   * `workerId` as key and an integer value
+   */
+  public readonly tasks: Map<Worker, number> = new Map<Worker, number>()
+
+  public readonly emitter: PoolEmitter
+
+  protected id: number = 0
+
+  public constructor (
+    public readonly numWorkers: number,
+    public readonly filePath: string,
+    public readonly opts: PoolOptions<Worker> = { maxTasks: 1000 }
+  ) {
+    if (!this.isMain()) {
+      throw new Error('Cannot start a pool from a worker!')
+    }
+    // TODO christopher 2021-02-07: Improve this check e.g. with a pattern or blank check
+    if (!this.filePath) {
+      throw new Error('Please specify a file with a worker implementation')
+    }
+
+    this.setupHook()
+
+    for (let i = 1; i <= this.numWorkers; i++) {
+      this.internalNewWorker()
+    }
+
+    this.emitter = new PoolEmitter()
+  }
+
+  protected setupHook (): void {
+    // Can be overridden
+  }
+
+  protected abstract isMain (): boolean
+
+  public async destroy (): Promise<void> {
+    for (const worker of this.workers) {
+      await this.destroyWorker(worker)
+    }
+  }
+
+  protected abstract destroyWorker (worker: Worker): void | Promise<void>
+
+  protected abstract sendToWorker (
+    worker: Worker,
+    message: MessageValue<Data>
+  ): void
+
+  /**
+   * Execute the task specified into the constructor with the data parameter.
+   *
+   * @param data The input for the task specified.
+   * @returns Promise that is resolved when the task is done.
+   */
+  public execute (data: Data): Promise<Response> {
+    // configure worker to handle message with the specified task
+    const worker = this.chooseWorker()
+    const previousWorkerIndex = this.tasks.get(worker)
+    if (previousWorkerIndex !== undefined) {
+      this.tasks.set(worker, previousWorkerIndex + 1)
+    } else {
+      throw Error('Worker could not be found in tasks map')
+    }
+    const id = ++this.id
+    const res = this.internalExecute(worker, id)
+    this.sendToWorker(worker, { data: data || ({} as Data), id: id })
+    return res
+  }
+
+  protected abstract registerWorkerMessageListener (
+    port: Worker,
+    listener: (message: MessageValue<Response>) => void
+  ): void
+
+  protected abstract unregisterWorkerMessageListener (
+    port: Worker,
+    listener: (message: MessageValue<Response>) => void
+  ): void
+
+  protected internalExecute (worker: Worker, id: number): Promise<Response> {
```

Comment:
As a C developer, I'm used to __execute() way of naming for function used internally. 
Is there a similar convention in the TS, JS world?

Reply:
In Vue/Vuetify they use `internal` as prefix
The underscore (`_`) prefix is the more oldish way of writing a private field/method
Nowadays we have `#` as a real private js field/method (it can really not be accessed from outside!)
But in typescript you have the `private` keyword, and so you don't need the `_` prefix

The `__` is one way to solve this, it's interchangeable with `internal`, but reads much more technical :shrug: 
@pioardi what would you prefer? `__` or `internal`?

Reply:
I would stay with the internal, we could add the # if it make the methods really private.
But is not mandatory for now in my opinion